### PR TITLE
Add Start/Stop Timer Commands

### LINKED ISSUE #9: As a new user, I can start and stop a timer for a project
... so that I can track the time I spent on the project.

### PR DESCRIPTION
**Pull request by boundtotheearth**
Closes #9 

More comprehensive and reliable tests need to be added.

### CODE REVIEW

File: src/main/java/seedu/momentum/logic/parser/StopCommandParser.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/logic/parser/StopCommandParser.java]: Javadoc here should be `StopCommand` instead of `StartCommand`

File: src/main/java/seedu/momentum/model/ModelManager.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/ModelManager.java]: Should these two checks be extracted out into another method to keep the `equals()` method here clean?

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/Project.java]: Should the javadoc here be more descriptive?

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/Project.java]: Should have an empty line before the `@param` tag

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/Project.java:123]: Should the `equals()` method below check whether both timers are equal?

File: src/main/java/seedu/momentum/model/timer/Timer.java
Comment (claracheong4) [inline src/main/java/seedu/momentum/model/timer/Timer.java]: Do we really need to store the entire project object in a timer?

**Review by claracheong4 — COMMENTED** [review]
I think everything looks alright except that a timer contains a project.
I thought that it was agreed that a project should contain a timer instead.

File: src/main/java/seedu/momentum/model/timer/Timer.java
Comment (boundtotheearth) [reply src/main/java/seedu/momentum/model/timer/Timer.java]: The Timer class is only used to keep track of the active timers. Once the timer is stopped, the data is stored under a WorkDuration class (which has not reference to project), and put in a list inside the corresponding project.

I implemented it this way because I didn't want to store incomplete timers in a project, since it might cause problems if the timer wasn't stopped properly. Its also easier to check for active timers this way (see activeTimer in ModelManager)

And i think it would be easier to extend to having multiple concurrent timers if we decide to do that (then each timer needs to know which project its keeping track of)

Just storing the index won't work because the filtered list might change while the timer is running.

**Review by pr4aveen — COMMENTED** [review]
Looks good overall. Just some minor things to look at before merging.

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [reply src/main/java/seedu/momentum/model/project/Project.java:123]: By below I mean in the `Project.java` class, can't really highlight it from here.

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (khoodehui) [reply src/main/java/seedu/momentum/model/project/Project.java]: Yup think it will better to describe that it will construct a new project, just like the javadocs in other classes.

File: src/main/java/seedu/momentum/model/timer/Time.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/model/timer/Time.java]: Will it be better to give an example of how the ISO8601 time format looks like in the error message? A user might not know what this format is.

File: src/main/java/seedu/momentum/model/timer/Timer.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/model/timer/Timer.java]: Should have a new line before the param.

File: src/main/java/seedu/momentum/model/timer/UniqueDurationList.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/model/timer/UniqueDurationList.java:123]: Think the persons here should be timers instead.

File: src/main/java/seedu/momentum/model/timer/Timer.java
Comment (claracheong4) [reply src/main/java/seedu/momentum/model/timer/Timer.java]: Maybe we generate and use an unique project id?

File: src/main/java/seedu/momentum/model/ModelManager.java
Comment (boundtotheearth) [reply src/main/java/seedu/momentum/model/ModelManager.java]: Actually I think the Optional.equals() method also checks if both are empty, so I can just remove the 1st line

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (boundtotheearth) [reply src/main/java/seedu/momentum/model/project/Project.java]: Haha the javadocs is from the original addressbook. I'll change it

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (boundtotheearth) [reply src/main/java/seedu/momentum/model/project/Project.java:123]: This is mostly what i was referring to when i said more/better tests are needed.

Basically, since I'm using LocalDateTime.now() to get the current time, 2 times will never be exactly the same (they will be some nanoseconds apart). For example in tests, even if I start 2 timers immediately after one another, they will have different values, so equals() will still say that they're different. Since the existing tests involve checking if 2 models are equal, this will break a bunch of existing tests.

I'm think of either (a) Overloading the timer functions so that I can start them at a fixed time (just for testing) or (b) Introduce a custom global clock for the application to run on. The global clock will be the same as system clock in production, but in testing it will have functions that allow us to manually control the time.

File: src/main/java/seedu/momentum/model/timer/Time.java
Comment (boundtotheearth) [reply src/main/java/seedu/momentum/model/timer/Time.java]: A user will never see this message (as of now) since we don't require users to manually input times. If we add commands that involve time inputs, then it might be better to not use ISO8601 format in the first place (its hard to read) and instead use some variation of the "yyyy MM dd hh:mm:ss" style.

This error only come up in logs when theres an error in loading saved data (e.g. the saved data is not in ISO format). For now I'll add an example of what the format looks like to the error message

File: src/test/java/seedu/momentum/logic/commands/StopCommandTest.java
Comment (kkangs0226) [inline src/test/java/seedu/momentum/logic/commands/StopCommandTest.java:60]: I think should have empty line here??

File: src/test/java/seedu/momentum/model/timer/WorkDurationTest.java
Comment (kkangs0226) [inline src/test/java/seedu/momentum/model/timer/WorkDurationTest.java:51]: Not sure if I understood your TypicalWorkDuration class fully, but is it comparing different projects here?

**Review by kkangs0226 — COMMENTED** [review]
LGTM other than the two comments!

File: src/test/java/seedu/momentum/model/timer/WorkDurationTest.java
Comment (boundtotheearth) [reply src/test/java/seedu/momentum/model/timer/WorkDurationTest.java:51]: TypicalWorkDuraiton is just a class to provide dummy data for testing. I defined some simple durations like 1 day, week, month, etc for testing

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/Project.java]: Javadoc comment needs to be updated to include timer

File: src/main/java/seedu/momentum/model/project/Project.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/project/Project.java]: Why not just use the `otherProject` variable here?

File: src/main/java/seedu/momentum/model/timer/Time.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/timer/Time.java]: There should be an extra line before `@params` tag and a description for what is returned after the `@return` tag

File: src/main/java/seedu/momentum/model/timer/Timer.java
Comment (pr4aveen) [inline src/main/java/seedu/momentum/model/timer/Timer.java:67]: Should include an empty line before the `@params` tag

File: src/test/java/seedu/momentum/model/project/ProjectTest.java
Comment (pr4aveen) [inline src/test/java/seedu/momentum/model/project/ProjectTest.java:60]: Can I just check what the reason for changing this was?

**Review by pr4aveen — APPROVED** [review]
Minor coding standard violations. Otherwise, LGTM 👍

File: src/main/java/seedu/momentum/commons/core/Clock.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/commons/core/Clock.java:41]: Should have a new line after the description. Same for the 2 methods above.

File: src/main/java/seedu/momentum/logic/parser/StopCommandParser.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/logic/parser/StopCommandParser.java:18]: Newline after description.

File: src/main/java/seedu/momentum/model/timer/Timer.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/model/timer/Timer.java]: Extra line here should be removed.

File: src/main/java/seedu/momentum/model/timer/WorkDuration.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/model/timer/WorkDuration.java]: Extra asterisk here should be removed.

File: src/main/java/seedu/momentum/model/timer/WorkDuration.java
Comment (khoodehui) [inline src/main/java/seedu/momentum/model/timer/WorkDuration.java:36]: Newline after description.

File: src/test/java/seedu/momentum/logic/commands/StartCommandTest.java
Comment (khoodehui) [inline src/test/java/seedu/momentum/logic/commands/StartCommandTest.java:58]: Newline before method.

**Review by khoodehui — APPROVED** [review]
LGTM after the coding standard fixes!

File: src/test/java/seedu/momentum/model/project/ProjectTest.java
Comment (boundtotheearth) [reply src/test/java/seedu/momentum/model/project/ProjectTest.java:60]: It was related to the immutability thing I told you about last time. Its not relevant anymore but they're both equivalent

**Pull request merged**
